File: src/main/java/com/chethan/assignment3/employee/EmployeeImplementation.java
package com.chethan.assignment3.employee;

import java.time.*;
import java.util.*;
import java.util.stream.*;

/**
 * Created by Chethan on Feb 16, 2022.
 */

public class EmployeeImplementation {


    //No of employees by given department
    public static NoOfEmpByDeptContainer getNoOfEmployeesByDept(List<Employee> employeeList, String dept) {

        return new NoOfEmpByDeptContainer(dept, employeeList.stream().filter(i -> i.getDepartment().equals(dept)).count());
    }


    //Group employees by department
    public static List<GroupEmployeesByDepartment> getGroupByDept(List<Employee> employeeList) {

        List<String>      distinctDept = employeeList.stream().map(i -> i.getDepartment()).distinct().collect(Collectors.toList());
        Map<String, Long> m1           = new TreeMap<String, Long>();
        return distinctDept.stream().map(i -> new GroupEmployeesByDepartment(i,
                employeeList.stream().filter(k -> i.equals(k.getDepartment())).count())).collect(Collectors.toList());
    }

    //Increase salary for given department
    public static Map<String, Double> getIncreasedSalary(List<Employee> employeeList, Optional<String> dept, Optional<Double> hike) {
        Map<String, Double> promotedEmployees = new TreeMap<String, Double>();
        if (dept.isPresent()) {
            employeeList.stream().filter(f -> f.getDepartment().equals (dept.get())).forEach(m -> { EmployeeDatabase.updateStudentDetails(m.getId(), "salary", ""+(m.getSalary()+m.getSalary()*(hike.get()/100)));
                promotedEmployees.put(m.getName(), m.getSalary()+m.getSalary()*(hike.get()/100));
            });
        }
        return promotedEmployees;
    }

    //Promote employees based on experience
    public static Map<String, String> getPromotedEmployees(List<Employee> employeeList) {
        Map<String, String>                             promotedEmployees = new TreeMap<String, String>();
        List<Employee> promoteList       = employeeList.stream().filter(f -> Period.between(f.getJoiningDate(), LocalDate.now()).getYears() >= 8).collect(Collectors.toList());
        promoteList.stream().forEach(i -> {EmployeeDatabase.updateStudentDetails(i.getId(),"jobLevel","senior");
            promotedEmployees.put(i.getName(), i.getJobLevel());
        });
        return promotedEmployees;
    }
}

File: src/main/scala/com/chethan/assignment1/EmployeeImplementation.scala
package com.chethan.assignment1

import java.time.{LocalDate, Period}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

/**
  * Created by Chethan on Jul 26, 2022.
  */

object EmployeeImplementation {
  val employeesListInFuture: Future[List[Employee]] = new EmployeeDatabase().getAllData

  def getNoOfEmployeesByDept(dept: String): Future[(String, Int)] =
    employeesListInFuture.map(listOfEmployees => (dept, listOfEmployees.count(employee => employee.department.equals(dept))))


  def groupEmployeesByDepartment: Future[List[EmployeesByDepartment]] = {
    for {
      employeeList <- employeesListInFuture
      employeesCount = employeeList.groupBy(employee => employee.department)
      employeesCount2 = employeeList.groupBy(employee => employee.department).map{ case (departmentName, employees) =>
        EmployeesByDepartment(departmentName, employees)
      }.toList
    } yield {
      //      EmployeesByDepartment("ec", employeesCount("ec")) +: EmployeesByDepartment("cs", employeesCount("cs")) +: EmployeesByDepartment("ee", employeesCount("ee")) +: Nil

      employeesCount2
    }
  }

  def increaseSalaryOfEmployees(department: String, hikeInPercentage: Option[Double]): Future[List[Boolean]] = {
//
//    for (employeeList <- employeesListInFuture)
//      yield if (hikeInPercentage.isDefined)
//              employeeList.filter(employee => employee.department.equals(department))
//                          .map(hikedEmployee => {
//                            val updatedEmployeeRecord = Employee(hikedEmployee.id, hikedEmployee.name, hikedEmployee.department, hikedEmployee.salary + (hikedEmployee.salary * (hikeInPercentage.get / 100)),
//                              hikedEmployee.gender, hikedEmployee.joiningDate, hikedEmployee.dob, hikedEmployee.jobLevel)
//                            new EmployeeDatabase().updateData(updatedEmployeeRecord).value.get.get // todo not recommended
//                          })
//
//            else throw new RuntimeException("hike not found exception")
    for {
      employeeList <- employeesListInFuture
      result: List[Boolean] <- if (hikeInPercentage.isEmpty) {
        Future.successful(Nil)
      } else {
        val filteredEmployees              = employeeList.filter(employee => employee.department.equals(department))
        val records: List[Future[Boolean]] = filteredEmployees.map{ employee =>
          val updatedSalary   = employee.salary + (employee.salary * (hikeInPercentage.get / 100))
          val updatedEmployee = employee.copy(salary = updatedSalary)
          new EmployeeDatabase().updateData(updatedEmployee)
        }
        Future.sequence(records)

      }
    } yield result
  }

  def promoteEmployees(): Future[List[Employee]] = {
    employeesListInFuture.map(listOfEmployees =>
      listOfEmployees.filter(employee => Period.between(employee.joiningDate, LocalDate.now()).getYears >= 8).map(promotedEmployee => {
        new EmployeeDatabase().updateData(Employee(promotedEmployee.id, promotedEmployee.name, promotedEmployee
          .department, promotedEmployee.salary, promotedEmployee.gender, promotedEmployee.joiningDate, promotedEmployee.dob, "senior"))
        new EmployeeDatabase().getDataById(promotedEmployee.id).value.get.get
      }))
  }
}


case class EmployeesByDepartment(department: String, listOfEmployees: List[Employee])
